### Шаг 1. Изучение общей информации

Открытие файла, просмотр таблицы:

In [1]:
import pandas as pd
import ipywidgets as widgets
from IPython.display import display
from pymystem3 import Mystem
import copy
m = Mystem()
df = pd.read_csv('datasets/data.csv')
df.head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
0           0         1   -8437.673028         42    высшее             0   
1           1         1   -4024.803754         36   среднее             1   
2           2         0   -5623.422610         33   Среднее             1   
3           3         3   -4124.747207         32   среднее             1   
4           4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Просмотр типов данных и количество пропусков:

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
Unnamed: 0          21525 non-null int64
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(6), object(5)
memory usage: 2.1+ MB


Просмотр имен столбцов:

In [3]:
df.columns

Index(['Unnamed: 0', 'children', 'days_employed', 'dob_years', 'education',
       'education_id', 'family_status', 'family_status_id', 'gender',
       'income_type', 'debt', 'total_income', 'purpose'],
      dtype='object')

Просмотр таблицы с сортировкой по стажу:

In [4]:
df.sort_values(by='days_employed').head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
16335       16335         1  -18388.949901         61   среднее             1   
4299         4299         0  -17615.563266         61   среднее             1   
7329         7329         0  -16593.472817         60    высшее             0   
17838       17838         0  -16264.699501         59   среднее             1   
16825       16825         0  -16119.687737         64   среднее             1   

         family_status  family_status_id gender income_type  debt  \
16335  женат / замужем                 0      F   сотрудник     0   
4299   женат / замужем                 0      F   компаньон     0   
7329   женат / замужем                 0      F   сотрудник     0   
17838  женат / замужем                 0      F   сотрудник     0   
16825  женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
16335  186178.934089      операции с недвижимостью  
4299   122560.741753                 покупка жилья  
7329   124697.846781  заняться высшим образованием  
17838   51238.967133         на покупку автомобиля  
16825   91527.685995    покупка жилой недвижимости

In [5]:
df.sort_values(by='days_employed', ascending=False).head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
6954         6954         0  401755.400475         56   среднее             1   
10006       10006         0  401715.811749         69    высшее             0   
7664         7664         1  401675.093434         61   среднее             1   
2156         2156         0  401674.466633         60   среднее             1   
7794         7794         0  401663.850046         61   среднее             1   

               family_status  family_status_id gender income_type  debt  \
6954          вдовец / вдова                 2      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7794        гражданский брак                 1      F   пенсионер     0   

        total_income                purpose  
6954   176278.441171           ремонт жилью  
10006   57390.256908  получение образования  
7664   126214.519212      операции с жильем  
2156   325395.724541             автомобили  
7794    48286.441362                свадьба

Просмотр колонки с детьми:

In [6]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Просмотр колонки с возрастом заемщика:

In [7]:
df['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Просмотр колонки с возрастом заемщика моложе 18 лет:

In [8]:
df.query('dob_years < 18').count()

Unnamed: 0          101
children            101
days_employed        91
dob_years           101
education           101
education_id        101
family_status       101
family_status_id    101
gender              101
income_type         101
debt                101
total_income         91
purpose             101
dtype: int64

Просмотр колонки с категорией образования:

In [9]:
df['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

Просмотр колонки с категорией семейного положения:

In [10]:
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Просмотр колонки с категорией просрочки:

In [11]:
df['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

Просмотр колонки с отрицательными доходами:

In [12]:
df.query('total_income < 0').count()

Unnamed: 0          0
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Вывод

В колонках о стаже и доходе присутствуют пропуски, которые нужно заполнить, и дробные значения, которые нужно привести к целым для оптимизации. Колонки имеют корректные названия. В колонке со стажем присутствуют неадекватные значения двух видов (в десятки раз завышенные и отрицательные). В колонке с детьми присутствуют неадекватные значения двух видов (-1 и 20). В колонке с возрастом заемщика присутствуют неадекватные нулевые данные. Причина появления неадекватных данных и пропусков - технический сбой.

### Шаг 2. Предобработка данных

### Обработка пропусков

Посчитаем количество пропусков в колонке со стажем:

In [13]:
df['days_employed'].isnull().value_counts()

False    19351
True      2174
Name: days_employed, dtype: int64

Посчитаем количество пропусков в колонке с доходом:

In [14]:
df['total_income'].isnull().value_counts()

False    19351
True      2174
Name: total_income, dtype: int64

2174 пропусков в колонках со стажем и доходом. Проверим, не одновременно ли встречаются пропуски в этих колонках:

In [15]:
df[(df['days_employed'].isnull()) & (df['total_income'].isnull())].head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
12          12         0            NaN         65   среднее             1   
26          26         0            NaN         41   среднее             1   
29          29         0            NaN         63   среднее             1   
41          41         0            NaN         50   среднее             1   
55          55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Пропуски встречаются одновременно в обеих колонках. Заполним пропуски медианными значениями колонку со стажем:

In [16]:
median = df['days_employed'].median()
df['days_employed'] = df['days_employed'].fillna(median)
df.query('days_employed == @median').head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
12          12         0   -1203.369529         65   среднее             1   
26          26         0   -1203.369529         41   среднее             1   
29          29         0   -1203.369529         63   среднее             1   
41          41         0   -1203.369529         50   среднее             1   
55          55         0   -1203.369529         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Заполним пропуски медианными значениями колонку с доходом:

In [17]:
median = df['total_income'].median()
df['total_income'] = df['total_income'].fillna(median)
df.query('total_income == @median').head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
12          12         0   -1203.369529         65   среднее             1   
26          26         0   -1203.369529         41   среднее             1   
29          29         0   -1203.369529         63   среднее             1   
41          41         0   -1203.369529         50   среднее             1   
55          55         0   -1203.369529         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

     total_income                           purpose  
12  145017.937533                   сыграть свадьбу  
26  145017.937533                       образование  
29  145017.937533  строительство жилой недвижимости  
41  145017.937533  сделка с подержанным автомобилем  
55  145017.937533                   сыграть свадьбу

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
Unnamed: 0          21525 non-null int64
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(6), object(5)
memory usage: 2.1+ MB


Пропуски отсутствуют.

### Вывод

Пропущены значения у 2174 строк, что составляет 10.1% от общего числа записей, одновременно в столбцах о стаже работы и месячном доходе. Банк, когда формирует заявку на выдачу кредита, обязательно запрашивает у клиента трудовой стаж и месячный доход. Следовательно, данные отсутствуют по технической причине. Пропущенные значения заполнены медианами.

### Замена типа данных

Неоптимальный тип данных присутствует опять же в колонках со стажем и доходом. Преобразуем их в целочисленный тип:

In [19]:
df = df.astype({'days_employed': 'int', 'total_income': 'int'})

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
Unnamed: 0          21525 non-null int64
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(8), object(5)
memory usage: 2.1+ MB


### Вывод

Вещественные типы имеются у колонок со стажем и доходом, что является избыточным - доли дней стажа и рублей дохода роли не играют, а только лишь затрудняют читабельность и производительность. Подлежат замене на целочисленный тип методом astype объекта DataFrame как наиболее подходящим в данной ситуации.

### Обработка дубликатов

Посмотрим полные дубли:

In [21]:
df.duplicated().sum()

0

Удаляем полные дубли:

In [22]:
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

Полные дубли удалены. Посмотрим на дубли в колонке с образованием:

In [23]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

В дублях есть отличие только по регистру. Преобразуем значения к нижнему регистру:

In [24]:
df['education'] = df['education'].str.lower()
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Дубли в колонке с образованием удалены. Посмотрим дубли в колонке с семеным положением:

In [25]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Хоть дублей и нет, желательно привести к одному регистру все записи в колонке с семейным положением:

In [26]:
df['family_status'] = df['family_status'].str.lower()
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Посмотри дубли в колонке с полом:

In [27]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Дублей нет, но есть странный пол "XNA". Наверное, что-то новомодное. ))
Посмотрим на дубли в колонке с типом занятости:

In [28]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

В колонке с типом занятости дублей нет. Посмотрим дубли в целях кредита:

In [29]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Здесь нет формальных дублей, есть смысловые дубли, и справится здесь лемматизация.

### Вывод

Выявлены 54 полных дубля методом duplicated() и удалены методом drop_duplicates() с перестройкой индекса reset_index(). Частичные (смысловые) дубли выявлялись в строковых данных методом value_counts() для отдельных колонок. Дубли по регистру преобразованы к нижнему регистру с помощью str.lower(). В колонке с целью кредита есть множество значений, с которыми нужно работать методом лемматизации, так как слова записаны в разных формах. Причины появления дублей вижу в отсутствии единых категорий при формировании заявки. Кредитный специалист заполняет заявку, вручную заполняя данные со слов клиента. Лучше предоставить кредитному специалисту единообразный список значений, из которого можно выбирать нужные.

### Лемматизация

Получим список лемм в колонке с целью кредита, преобразуем получившийся список в строку, удалим из строки концевые пробельные символы:

In [30]:
df['purpose'] = df['purpose'].apply(m.lemmatize).str.join(sep='').str.rstrip()
df['purpose'].value_counts()

автомобиль                                973
свадьба                                   797
на проведение свадьба                     777
сыграть свадьба                           774
операция с недвижимость                   676
покупка коммерческий недвижимость         664
операция с жилье                          653
покупка жилье для сдача                   653
операция с коммерческий недвижимость      651
покупка жилье                             647
жилье                                     647
покупка жилье для семья                   641
строительство собственный недвижимость    635
недвижимость                              634
операция со свой недвижимость             630
строительство жилой недвижимость          626
покупка недвижимость                      624
строительство недвижимость                620
покупка свой жилье                        620
ремонт жилье                              612
покупка жилой недвижимость                607
на покупка свой автомобиль        

### Вывод

Строки в колонке с целью кредита преобразованы в леммы.

### Категоризация данных

Создадим функцию по обработке лемм, которая возвращает стандартные значения для целей кредита:

In [31]:
def purpose_category(row):
    if 'автомобиль' in row:
        return 'автомобиль'
    if 'свадьба' in row:
        return 'свадьба'
    if 'недвижимость' in row:
        if 'коммерческий' in row:
            return 'покупка коммерческой недвижимости'
        if 'строительство' in row:
            return 'строительство'
        return 'покупка жилой недвижимости'
    if 'жилье' in row:
        if 'ремонт' in row:
            return 'ремонт жилой недвижимости'
        return 'покупка жилой недвижимости'
    if 'образование' in row:
        return 'образование'
    return 'прочее'

Создадим на основе функции колонку со стандартными целями кредита:

In [32]:
df['purpose_standart'] = df['purpose'].apply(purpose_category)
df['purpose_standart'].value_counts()

покупка жилой недвижимости           7032
автомобиль                           4315
образование                          4022
свадьба                              2348
строительство                        1881
покупка коммерческой недвижимости    1315
ремонт жилой недвижимости             612
Name: purpose_standart, dtype: int64

Удалим ставшую ненужной колонку с необработанными целями кредита:

In [33]:
df = df.drop(columns='purpose')

Посмотрим, какие еще нужно выделить категории:

In [34]:
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

In [35]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [36]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [37]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

In [38]:
df['purpose_standart'].value_counts()

покупка жилой недвижимости           7032
автомобиль                           4315
образование                          4022
свадьба                              2348
строительство                        1881
покупка коммерческой недвижимости    1315
ремонт жилой недвижимости             612
Name: purpose_standart, dtype: int64

Для категоризации подходят колонки с образованием, семейным положением, типом занятостии, целью кредита, так как они содержат длинные строки. Выделим категорию с образованием в отдельную таблицу:

In [39]:
education_dict = df[['education_id', 'education']].drop_duplicates().reset_index(drop=True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Выделим категорию с полом в отдельную таблицу для будущего использования:

In [40]:
gender_dict = df[['gender']].drop_duplicates().reset_index(drop=True)
gender_dict

gender
0      F
1      M
2    XNA

Выделим категорию с семейным положением в отдельную таблицу:

In [41]:
family_status_dict = df[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

Выделим категорию с типом занятости в отдельную таблицу, в которой также создадим колонку с номером:

In [42]:
income_type_dict = df[['income_type']].drop_duplicates().reset_index(drop=True)
income_type_dict['income_type_id'] = range(income_type_dict.shape[0])
income_type_dict

income_type  income_type_id
0        сотрудник               0
1        пенсионер               1
2        компаньон               2
3      госслужащий               3
4      безработный               4
5  предприниматель               5
6          студент               6
7        в декрете               7

Добавим колонку с номером категории из словаря с типом занятости в общую таблицу, чтобы в дальнешем можно было обращаться к категории с типом занятости в общей таблице по индексу:

In [43]:
df = df.merge(income_type_dict, on='income_type', how='left')
df.head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
0           0         1          -8437         42    высшее             0   
1           1         1          -4024         36   среднее             1   
2           2         0          -5623         33   среднее             1   
3           3         3          -4124         32   среднее             1   
4           4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

             purpose_standart  income_type_id  
0  покупка жилой недвижимости               0  
1                  автомобиль               0  
2  покупка жилой недвижимости               0  
3                 образование               0  
4                     свадьба               1

Выделим категорию с целью кредита в отдельную таблицу, в которой также создадим колонку с номером:

In [44]:
purpose_dict = df[['purpose_standart']].drop_duplicates().reset_index(drop=True)
purpose_dict['purpose_standart_id'] = range(purpose_dict.shape[0])
purpose_dict

purpose_standart  purpose_standart_id
0         покупка жилой недвижимости                    0
1                         автомобиль                    1
2                        образование                    2
3                            свадьба                    3
4  покупка коммерческой недвижимости                    4
5                      строительство                    5
6          ремонт жилой недвижимости                    6

Добавим колонку с номером категории из словаря с целями кредита в общую таблицу, чтобы в дальнешем можно было обращаться к категории с целью кредита в общей таблице по индексу:

In [45]:
df = df.merge(purpose_dict, on='purpose_standart', how='left')
df.head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
0           0         1          -8437         42    высшее             0   
1           1         1          -4024         36   среднее             1   
2           2         0          -5623         33   среднее             1   
3           3         3          -4124         32   среднее             1   
4           4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

             purpose_standart  income_type_id  purpose_standart_id  
0  покупка жилой недвижимости               0                    0  
1                  автомобиль               0                    1  
2  покупка жилой недвижимости               0                    0  
3                 образование               0                    2  
4                     свадьба               1                    3

Создадим категорию платежеспособности клиентов, для чего разделим их на 5 групп по уровню доходов. Сначала создадим функцию, возвращающую номер категории, от 0 до 5. Чем выше номер категории, тем выше доход:

In [46]:
def income_category(value):
    if value == 0:
        return 0
    levels = 5
    for level in range(levels):
        if value <= (df['total_income'].max() / levels * (level + 1)):
            return (level + 1)
    return levels

Создадим колонку с номером категории доходов клиента в основной таблице:

In [47]:
df['income_category'] = df['total_income'].apply(income_category).astype('int')

Создадим словарь с категориями доходов клиентов:

In [48]:
income_dict = df.sort_values(by='income_category')[['income_category']].drop_duplicates().reset_index(drop=True)
income_dict

income_category
0                1
1                2
2                3
3                4
4                5

Создадим функцию для добавления диапазона доходов в словарь с категориями доходов клиентов:

In [49]:
def income_range(category):
    if category == 0:
        return 0
    levels = 5
    low = int(df['total_income'].max() / levels * (category - 1))
    high = int(df['total_income'].max() / levels * category)
    return f'{low} - {high}'

In [50]:
income_dict['income_range'] = income_dict['income_category'].apply(income_range)
income_dict

income_category       income_range
0                1         0 - 453120
1                2    453120 - 906241
2                3   906241 - 1359362
3                4  1359362 - 1812483
4                5  1812483 - 2265604

In [51]:
df.head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
0           0         1          -8437         42    высшее             0   
1           1         1          -4024         36   среднее             1   
2           2         0          -5623         33   среднее             1   
3           3         3          -4124         32   среднее             1   
4           4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

             purpose_standart  income_type_id  purpose_standart_id  \
0  покупка жилой недвижимости               0                    0   
1                  автомобиль               0                    1   
2  покупка жилой недвижимости               0                    0   
3                 образование               0                    2   
4                     свадьба               1                    3   

   income_category  
0                1  
1                1  
2                1  
3                1  
4                1

Создание 5 групп доходов клиентов с помощью квантилей, предварительно полученных в другом блокноте:

In [52]:
#bins=jenkspy.jenks_breaks(df['total_income'], nb_class=5)
bins = [20667, 123508, 209228, 357169, 786451, 2265604]
df['income_group'] = pd.cut(
    df['total_income'],
    bins=bins,
    labels=['1', '2', '3', '4', '5'],
    include_lowest=True)
df.sort_values(by='total_income').head()

Unnamed: 0  children  days_employed  dob_years education  education_id  \
14585       14585         0         359219         57   среднее             1   
13006       13006         0         369708         37   среднее             1   
16174       16174         1          -3642         52   среднее             1   
1598         1598         0         359726         68   среднее             1   
14276       14276         0         346602         61   среднее             1   

          family_status  family_status_id gender income_type  debt  \
14585   женат / замужем                 0      F   пенсионер     1   
13006  гражданский брак                 1      M   пенсионер     0   
16174   женат / замужем                 0      M   сотрудник     0   
1598   гражданский брак                 1      M   пенсионер     0   
14276   женат / замужем                 0      F   пенсионер     0   

       total_income            purpose_standart  income_type_id  \
14585         20667  покупка жилой недвижимости               1   
13006         21205                 образование               1   
16174         21367                  автомобиль               0   
1598          21695                     свадьба               1   
14276         21895  покупка жилой недвижимости               1   

       purpose_standart_id  income_category income_group  
14585                    0                1            1  
13006                    2                1            1  
16174                    1                1            1  
1598                     3                1            1  
14276                    0                1            1

### Вывод

Леммы преобразованы в стандартные цели кредитов для удобства работы. Выделены категории в отдельные таблицы со строковым типами: образование, семейное положение, тип занятости, цель кредита. Также созданы группы клиентов по доходу и помещены в отдельный словарь с указанием диапазона дохода. Не выделен в категорию пол заемщика, так как в нем три простых значения (F, M, XNA), не нуждающиеся в категоризации. В общую таблицу добавлены индексы категорий с типом занятости, целью кредита и доходом.

### Шаг 3. Ответы на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Создадим сводную таблицу с категориями клиентов по количеству детей, вычислим доли плательщиков, удалим лишние колонки и строку:

In [53]:
pivot_children = df.pivot_table(index='children', columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['days_employed']
pivot_children['debt_no'] = round(pivot_children[0] / pivot_children['total_clients'] * 100, 2)
pivot_children['debt_yes'] = round(pivot_children[1] / pivot_children['total_clients'] * 100, 2)
pivot_children = pivot_children.drop(labels=[0, 1], axis=1).drop(labels=-1, axis=0)
pivot_children

debt           total_clients  debt_no  debt_yes
children                                       
0                      14149    92.49      7.51
1                       4818    90.78      9.22
2                       2055    90.56      9.44
3                        330    91.82      8.18
4                         41    90.24      9.76
5                          9   100.00      0.00
20                        76    89.47     10.53
total_clients          21525    91.91      8.09

In [54]:
# Вот еще вариант
temp = df.pivot_table(index = 'children', aggfunc = {'debt': ['mean', 'count']})
temp['debt_yes'] = round(temp['debt', 'mean'] * 100, 2)
temp['total_clients'] = temp['debt', 'count']
temp = temp.drop(labels=-1, axis=0).drop(labels=['debt'], axis=1) 
temp.columns

MultiIndex([(     'debt_yes', ''),
            ('total_clients', '')],
           )

### Вывод

В приведенной таблице видно, что больше всего заемщиков без детей, и платят они исправнее, чем заемщики с детьми. Исключение составляют заемщики с пятью детьми, все они платят в срок. Больше всего просрочек у родителей-героев, имеющих 20 детей (если это не техническая ошибка), они допускают просрочки в 10.53% случаях.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

Создадим сводную таблицу, вычислим доли плательщиков в процентах:

In [55]:
data_pivot = df.pivot_table(index='family_status', columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['days_employed']
data_pivot['debt_no'] = round(data_pivot[0] / data_pivot['total_clients'] * 100, 2)
data_pivot['debt_yes'] = round(data_pivot[1] / data_pivot['total_clients'] * 100, 2)
data_pivot = data_pivot.drop(labels=[0, 1], axis=1)
data_pivot

debt                   total_clients  debt_no  debt_yes
family_status                                          
в разводе                       1195    92.89      7.11
вдовец / вдова                   960    93.44      6.56
гражданский брак                4177    90.71      9.29
женат / замужем                12380    92.48      7.52
не женат / не замужем           2813    90.26      9.74
total_clients                  21525    91.91      8.09

### Вывод

Реже всех допускают просрочки вдовы и вдовцы (6.5?%). Чаще других допускают просрочки неженатые / незамужние (9.75%) и в гражданском браке (9.32%). Следовательно, официально регистрировавшие свои семейные отношения ответственней относятся к заемным деньгам.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Выше клиенты были равномерно разделены на 5 групп по доходам. Создадим сводную таблицу с категориями клиентов по уровню дохода, объединим ее со словарем с диапазоном доходов:

In [56]:
data_pivot = df.pivot_table(index='income_category', columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['children']
data_pivot = data_pivot.merge(income_dict, on='income_category', how='left')

Вычислим доли плательщиков в срок и просрочками, удалим лишние колонки:

In [57]:
data_pivot['debt_no'] = data_pivot[0] / data_pivot['total_clients'] * 100
data_pivot['debt_yes'] = data_pivot[1] / data_pivot['total_clients'] * 100
data_pivot = data_pivot.drop(labels=[0, 1], axis=1)
data_pivot

income_category  total_clients       income_range     debt_no   debt_yes
0               1          21202         0 - 453120   91.873408   8.126592
1               2            293    453120 - 906241   94.539249   5.460751
2               3             22   906241 - 1359362   95.454545   4.545455
3               4              6  1359362 - 1812483  100.000000   0.000000
4               5              2  1812483 - 2265604   50.000000  50.000000
5   total_clients          21525                NaN   91.911731   8.088269

С помощью квантилей Дженкса:

In [58]:
data_pivot = df.pivot_table(index='income_group', columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['children']

income_range_list = copy.copy(bins) 
income_range_list.append(0)

data_pivot['income_top'] = income_range_list[1:]
data_pivot['debt_no'] = data_pivot[0] / data_pivot['total_clients'] * 100
data_pivot['debt_yes'] = round(data_pivot[1] / data_pivot['total_clients'] * 100, 2)
data_pivot = data_pivot.drop(labels=[0, 1], axis=1)
data_pivot

debt           total_clients  income_top    debt_no  debt_yes
income_group                                                 
1                       7214      123508  91.863044      8.14
2                       9763      209228  91.467787      8.53
3                       3745      357169  92.790387      7.21
4                        760      786451  93.815789      6.18
5                         43     2265604  90.697674      9.30
total_clients          21525           0  91.911731      8.09

### Вывод

В последней колонке debt_yes указана доля клиентов с просрочками для каждой группы доходов. С ростом доходов количество заемщиков уменьшается. Менее обеспеченные заемщики в среднем допускают больше просрочек платежей. Самые дисциплинированные плательщики в группе 4 без просрочек. Из двух богатейших клиентов в высшей категории один платил не во время.

- Как разные цели кредита влияют на его возврат в срок?

Тот же программный код, что и выше, просто заменена одна колонка:

In [59]:
data_pivot = df.pivot_table(index='purpose_standart', columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['children']
data_pivot['debt_no'] = data_pivot[0] / data_pivot['total_clients'] * 100
data_pivot['debt_yes'] = data_pivot[1] / data_pivot['total_clients'] * 100
data_pivot = data_pivot.drop(labels=[0, 1], axis=1)
data_pivot

debt                               total_clients    debt_no  debt_yes
purpose_standart                                                     
автомобиль                                  4315  90.660487  9.339513
образование                                 4022  90.800597  9.199403
покупка жилой недвижимости                  7032  92.832765  7.167235
покупка коммерческой недвижимости           1315  92.471483  7.528517
ремонт жилой недвижимости                    612  94.281046  5.718954
свадьба                                     2348  92.078365  7.921635
строительство                               1881  92.344498  7.655502
total_clients                              21525  91.911731  8.088269

### Вывод

Самые исправные плательщики те, кто ремонтирует жилье, с долей просрочек 5.77%. Чаще других пропускают платежи покупатели автомобиля с долей просрочек 9.35% и обучающиеся с просрочками в 9.22% случаях.

### Шаг 4. Общий вывод

Предоставленные выше выводы дадут банку больше информации по статистике просрочек, что позволит перераспределить выдаваемые кредиты в пользу тех, кто исправнее платит.

## Эксперименты для души

Многоуровневая детализация:

In [60]:
data_pivot = df.pivot_table(index=['children', 'education'], columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['education_id']
data_pivot['debt_no'] = round(data_pivot[0] / data_pivot['total_clients'] * 100, 2)
data_pivot['debt_yes'] = round(data_pivot[1] / data_pivot['total_clients'] * 100, 2)
data_pivot = data_pivot.drop(labels=[0, 1], axis=1).drop(labels=-1, level=0, axis=0)
data_pivot

debt                               total_clients  debt_no  debt_yes
children      education                                            
0             высшее                        3264    94.70      5.30
              начальное                      189    92.06      7.94
              неоконченное высшее            486    92.39      7.61
              среднее                      10206    91.79      8.21
              ученая степень                   4   100.00      0.00
1             высшее                        1325    94.57      5.43
              начальное                       60    83.33     16.67
              неоконченное высшее            175    88.57     11.43
              среднее                       3257    89.50     10.50
              ученая степень                   1   100.00      0.00
2             высшее                         556    94.78      5.22
              начальное                       25    76.00     24.00
              неоконченное высшее             72    87.50     12.50
              среднее                       1402    89.30     10.70
3             высшее                          83    95.18      4.82
              начальное                        7   100.00      0.00
              неоконченное высшее              8    75.00     25.00
              среднее                        231    90.91      9.09
              ученая степень                   1   100.00      0.00
4             высшее                           9   100.00      0.00
              среднее                         32    87.50     12.50
5             начальное                        1   100.00      0.00
              среднее                          8   100.00      0.00
20            высшее                          14   100.00      0.00
              неоконченное высшее              2   100.00      0.00
              среднее                         60    86.67     13.33
total_clients                              21525    91.91      8.09

In [61]:
data_pivot = df.pivot_table(index=['education', 'family_status', 'gender'], columns='debt', aggfunc='count', margins=True,
                            margins_name='total_clients', fill_value=0)['education_id']
data_pivot['debt_no'] = round(data_pivot[0] / data_pivot['total_clients'] * 100, 2)
data_pivot['debt_yes'] = round(data_pivot[1] / data_pivot['total_clients'] * 100, 2)
data_pivot = data_pivot.drop(labels=[0, 1], axis=1)
data_pivot

debt                                              total_clients  debt_no  \
education           family_status         gender                           
высшее              в разводе             F                 253    95.65   
                                          M                  63    93.65   
                    вдовец / вдова        F                 124    95.97   
                                          M                  12    83.33   
                    гражданский брак      F                 688    94.91   
                                          M                 268    92.54   
                    женат / замужем       F                1959    95.05   
                                          M                1087    94.66   
                    не женат / не замужем F                 523    96.37   
                                          M                 283    90.46   
начальное           в разводе             F                  15    86.67   
                                          M                   5   100.00   
                    вдовец / вдова        F                  27   100.00   
                                          M                   4   100.00   
                    гражданский брак      F                  40    82.50   
                                          M                  16    75.00   
                    женат / замужем       F                  68    91.18   
                                          M                  73    87.67   
                    не женат / не замужем F                  13    92.31   
                                          M                  21    90.48   
неоконченное высшее в разводе             F                  22    86.36   
                                          M                   9    88.89   
                    вдовец / вдова        F                  13    84.62   
                    гражданский брак      F                 108    90.74   
                                          M                  57    92.98   
                                          XNA                 1   100.00   
                    женат / замужем       F                 215    91.63   
                                          M                 126    90.48   
                    не женат / не замужем F                 102    92.16   
                                          M                  91    89.01   
среднее             в разводе             F                 646    93.03   
                                          M                 181    89.50   
                    вдовец / вдова        F                 741    93.93   
                                          M                  39    76.92   
                    гражданский брак      F                2032    91.09   
                                          M                 967    86.87   
                    женат / замужем       F                5551    92.63   
                                          M                3297    90.23   
                    не женат / не замужем F                1094    91.77   
                                          M                 685    82.92   
ученая степень      в разводе             M                   1   100.00   
                    женат / замужем       F                   2   100.00   
                                          M                   2   100.00   
                    не женат / не замужем M                   1   100.00   
total_clients                                             21525    91.91   

debt                                              debt_yes  
education           family_status         gender            
высшее              в разводе             F           4.35  
                                          M           6.35  
                    вдовец / вдова        F           4.03  
                                          M          16.67  
                    гражданский брак      F           5.09  


Инициализаци виджетов:

In [62]:
out = widgets.Output()
dropdown_children = widgets.Dropdown(options=pivot_children.drop(labels='total_clients', axis=0).index, 
                                     description='Дети:',)
dropdown_education = widgets.Dropdown(options=education_dict['education'], 
                                     description='Образование:',)
dropdown_family_status = widgets.Dropdown(options=family_status_dict['family_status'], 
                                     description='Семейное положение:',)
dropdown_gender = widgets.Dropdown(options=gender_dict['gender'], 
                                     description='Пол:',)
dropdown_income_type = widgets.Dropdown(options=income_type_dict['income_type'], 
                                     description='Тип занятости:',)
dropdown_purpose_standart = widgets.Dropdown(options=purpose_dict['purpose_standart'], 
                                     description='Цель кредита:',)
button_rates = widgets.Button(description="Рейтинг просрочек")
text_filter_data = widgets.IntText(value=1, description='Фильтр записей:', disabled=False)
text_filter_miss_statistics = widgets.IntText(value=1, description='Фильтр отсутствующих критериев:', disabled=False)
hbox1 = widgets.HBox([dropdown_children, dropdown_education])
hbox2 = widgets.HBox([dropdown_family_status, dropdown_gender])
hbox3 = widgets.HBox([dropdown_income_type, dropdown_purpose_standart])
hbox4 = widgets.HBox([text_filter_data, text_filter_miss_statistics])
hbox5 = widgets.HBox([button_rates])
box1 = widgets.Box([out])

Процедура вывода таблицы всех возможных просрочек:

In [63]:
def print_rates(button):
    out.clear_output()
    miss_statistics = 0
    client = {'children': dropdown_children.value,
              'education': dropdown_education.value,
              'family_status': dropdown_family_status.value,
              'gender': dropdown_gender.value,
              'income_type': dropdown_income_type.value,
              'purpose_standart': dropdown_purpose_standart.value,
             }
    big_table = pd.DataFrame()
    checked_parametrs = []
    for parametr1 in client:
        for parametr2 in client:
            if parametr1 != parametr2:
                if [parametr1, parametr2] not in checked_parametrs and [parametr2, parametr1] not in checked_parametrs:
                    checked_parametrs.append([parametr1, parametr2])
                    client_df = df[(df[parametr1] == client[parametr1]) & (df[parametr2] == client[parametr2])]
                    try:
                        data_pivot = client_df.pivot_table(index=[parametr1, parametr2], columns='debt', aggfunc='count', 
                                                           margins=True, margins_name='total_clients', 
                                                           fill_value=0)['dob_years']
                        data_pivot_rows = data_pivot['total_clients'].max()
                        if data_pivot_rows >= text_filter_data.value:
                            if 1 not in data_pivot.columns:
                                data_pivot[1] = 0
                            data_pivot['debt_yes'] = round(data_pivot[1] / data_pivot['total_clients'] * 100, 2)
                            data_pivot = data_pivot.drop(labels=[0, 1], axis=1).drop(labels='total_clients', axis=0)
                            big_table = big_table.append(data_pivot)
                        else:
                            miss_statistics += 1
                            with out:
                                print(f'Мало записей для {parametr1} и {parametr2}, всего {data_pivot_rows}')
                    except:
                        miss_statistics += 1
                        with out:
                            print(f'Отсутствует статистика для {parametr1} и {parametr2}')
    big_table.index.names = ['parametr1', 'parametr2']
    big_table.columns.names = ['']
    if miss_statistics < text_filter_miss_statistics.value:
        rate = round(big_table['debt_yes'].mean(), 2)
    else:
        rate = 'невычисляемая из-за нехватки данных'
    with out:
        display(big_table)
        print(f'Средняя просрочка таких клиентов {rate}')

Вывод виджетов:

In [64]:
button_rates.on_click(print_rates)
out.clear_output()
display(hbox1)
display(hbox2)
display(hbox3)
display(hbox4)
display(hbox5)
display(box1)

Box(children=(Output(),))

In [65]:
# пример работы обработчика события дропдауна
w = widgets.Dropdown(
    options=['Addition', 'Multiplication', 'Subtraction'],
    value='Addition',
    description='Task:',
)
def on_change(change):
    if change['type'] == 'change' and change['name'] == 'value':
        print ("changed to %s" % change['new'])
w.observe(on_change)
display(w)

Dropdown(description='Task:', options=('Addition', 'Multiplication', 'Subtraction'), value='Addition')